ebs/ebs.go

```diff
@@ -348,51 +364,36 @@ func (d *Driver) MarkVolumeForGC(volume *ec2.Volume) error {
 	return nil
 }
 
-func (d *Driver) CreateAndBuildFromSnapshot(oldVolume *ec2.Volume, snapshotVolume *ec2.Volume, opts map[string]string) (string, int64, error) {
-	// If the volume I would snapshot off of is the same as the one that I need to create then just return
-	if oldVolume != nil && *oldVolume.VolumeId == *snapshotVolume.VolumeId {
-		return *oldVolume.VolumeId, *oldVolume.Size * GB, nil
-	}
+func (d *Driver) CreateAndBuildFromSnapshot(volume *ec2.Volume, args *BuildArgs) (string, int64, error) {
 	// Create snapshot from volume in other AZ
-	snapshotId, err := d.ebsService.LaunchSnapshot(*snapshotVolume.VolumeId, fmt.Sprintf("Creating snaphot from volume=%v", *snapshotVolume.VolumeId), convertEc2TagsToMap(snapshotVolume.Tags))
+	snapshotId, err := d.ebsService.LaunchSnapshot(*volume.VolumeId, fmt.Sprintf("Creating snapshot from volume=%v", *volume.VolumeId), convertEc2TagsToMap(volume.Tags))
 	if err != nil {
 		return "", -1, err
 	}
-
 	snapshot, err := d.ebsService.GetSnapshot(snapshotId)
 	if err != nil {
 		return "", -1, err
 	}
-	return d.BuildFromSnapshot(oldVolume, snapshot, opts)
+	return d.BuildFromSnapshot(snapshot, args)
 }
 
-func (d *Driver) BuildFromSnapshot(oldVolume *ec2.Volume, snapshot *ec2.Snapshot, opts map[string]string) (string, int64, error) {
-	// If the snapshot to rebuild off of was created from the oldVolume then update tags and move on
-	if oldVolume != nil && *oldVolume.VolumeId == *snapshot.VolumeId {
-		return *oldVolume.VolumeId, *oldVolume.Size * GB, nil
-	}
-
-	// If there is an old volume then we will mark is for GarbageCollection
-	if err := d.MarkVolumeForGC(oldVolume); err != nil {
-		return "", -1, err
-	}
-
+func (d *Driver) BuildFromSnapshot(snapshot *ec2.Snapshot, args *BuildArgs) (string, int64, error) {
 	if err := d.ebsService.WaitForSnapshotComplete(*snapshot.SnapshotId); err != nil {
 		return "", -1, err
 	}
 	log.Debugf("Snapshot=%v is ready", *snapshot.SnapshotId)
 	snapshotVolumeSize := *snapshot.VolumeSize * GB
-	volumeSize, err := d.getSize(opts, snapshotVolumeSize)
+	volumeSize, err := d.getSize(args.opts, snapshotVolumeSize)
 	if err != nil {
-		return "", volumeSize, err
+		return "", -1, err
```

Comment:
if you are using BuildReturn type, the size might default to zero (i.e. zero value for int). AWS does not accept volumes of size 0 so we might be fine with that over -1. Thoughts?